<h1 align = 'center'><B><U>CSCN8000</U> - <U>Artificial Intelligence Algorithms and Mathematics</U></B></h1>
<h2 align = 'center'><B><U>Assignment</U> - <U>1</U></h2>

<h3 align = 'center'><B><I>Waseem Iqbal Shaikh (8902834)</I><B></h3>

### Problem Statement: Given the symptoms exhibited in a patient and a database of anonymized patient records, predict whether the patient is likely to have an illness (Parkinson’s disease). A model of this decision problem could be used by a program to provide decision support to medical professionals.

### Importing the necessary packages:

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import matplotlib as mpl
import matplotlib.pyplot as plt

### Loading the 'pd_speech_features.csv' dataset in a dataframe:

In [3]:
dataFrame = pd.read_csv("./pd_speech_features.csv")
dataFrame.head(10)

Unnamed: 0 Unnamed: 1 Baseline Features Unnamed: 3 Unnamed: 4 Unnamed: 5   
0         id     gender               PPE        DFA       RPDE  numPulses  \
1          0          1           0.85247    0.71826    0.57227        240   
2          0          1           0.76686    0.69481    0.53966        234   
3          0          1           0.85083    0.67604    0.58982        232   
4          1          0           0.41121    0.79672    0.59257        178   
5          1          0            0.3279    0.79782    0.53028        236   
6          1          0            0.5078    0.78744    0.65451        226   
7          2          1           0.76095    0.62145    0.54543        322   
8          2          1           0.83671    0.62079    0.51179        318   
9          2          1           0.80826    0.61766    0.50447        318   

         Unnamed: 6        Unnamed: 7          Unnamed: 8    Unnamed: 9  ...   
0  numPeriodsPulses  meanPeriodPulses  stdDevPeriodPulses  locPctJitter  ...  \
1               239        0.00806353            8.68E-05       0.00218  ...   
2               233       0.008258256            7.31E-05       0.00195  ...   
3               231        0.00833959            6.04E-05       0.00176  ...   
4               177       0.010857733         0.000182739       0.00419  ...   
5               235       0.008161574         0.002668863       0.00535  ...   
6               221       0.007631204         0.002696381       0.00783  ...   
7               321       0.005990989         0.000107266       0.00222  ...   
8               317       0.006073855         0.000135739       0.00282  ...   
9               317       0.006057188            6.93E-05       0.00161  ...   

                Unnamed: 745               Unnamed: 746   
0  tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
1                      1.562                     2.6445   
2                     1.5589                     3.6107   
3                     1.5643                     2.3308   
4                     3.7805                     3.5664   
5                     6.1727                     5.8416   
6                     4.8025                     5.0734   
7                   117.2678                    75.3156   
8                     3.8564                    11.8909   
9                      2.264                     6.3993   

                Unnamed: 747               Unnamed: 748   
0  tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
1                     3.8686                     4.2105   
2                    23.5155                    14.1962   
3                     9.4959                    10.7458   
4                     5.2558                    14.0403   
5                     6.0805                     5.7621   
6                     7.0166                     5.9966   
7                    32.0478                      7.706   
8                     7.2891                     4.3682   
9                     4.4165                     4.2662   

                Unnamed: 749               Unnamed: 750   
0  tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
1                     5.1221                     4.4625   
2                    11.0261                     9.5082   
3                    11.0177                     4.8066   
4                     4.2235                     4.6857   
5                     7.7817                    11.6891   
6                     5.2065                     7.4246   
7                      3.106                     4.6206   
8                     3.6443                      5.961   
9                     3.6357                     3.7346   

                Unnamed: 751               Unnamed: 752   
0  tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
1                     2.6202                     3.0004   
2                     6.5245                     6.3431   
3                     2.9199                     3.1495   
4                

### Data Analysis:

In [4]:
dataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 757 entries, 0 to 756
Columns: 755 entries, Unnamed: 0 to Unnamed: 754
dtypes: object(755)
memory usage: 4.4+ MB


- Checking for missing value:

In [5]:
print(dataFrame.isnull().sum())

Unnamed: 0           0
Unnamed: 1           0
Baseline Features    0
Unnamed: 3           0
Unnamed: 4           0
                    ..
Unnamed: 750         0
Unnamed: 751         0
Unnamed: 752         0
Unnamed: 753         0
Unnamed: 754         0
Length: 755, dtype: int64


- Checking for duplicate data:

In [6]:
print(f"The number of duplicate data: {dataFrame.duplicated().sum()}")

The number of duplicate data: 1


### After performing Data Analysis we can conlude:
- The data used in this study were gathered from 188 patients with PD (107 men and 81 women) with ages ranging from 33 to 87 (65.1±10.9) at the Department of Neurology in Cerrahpaşa Faculty of Medicine, Istanbul University. 
- The control group consists of 64 healthy individuals (23 men and 41 women) with ages varying between 41 and 82 (61.1±8.9). 
- During the data collection process, the microphone is set to 44.1 KHz and following the physician’s examination, the sustained phonation of the vowel /a/ was collected from each subject with three repetitions.
- There are no missing values in the dataset.
- There is only one row of duplicate data.
- The are 756 instances and 754 attributes.
- The datatype of the data is Integer.

### Data Preprocessing:

- As we can see, the first row of the data are feature names. Therefore we will change the coloumn name to feature names and drop the first row from the data:

In [7]:
dataFrame.columns = dataFrame.iloc[0]
dataFrame.drop(index = 0, inplace = True)
dataFrame.head()

0 id gender      PPE      DFA     RPDE numPulses numPeriodsPulses   
1  0      1  0.85247  0.71826  0.57227       240              239  \
2  0      1  0.76686  0.69481  0.53966       234              233   
3  0      1  0.85083  0.67604  0.58982       232              231   
4  1      0  0.41121  0.79672  0.59257       178              177   
5  1      0   0.3279  0.79782  0.53028       236              235   

0 meanPeriodPulses stdDevPeriodPulses locPctJitter  ...   
1       0.00806353           8.68E-05      0.00218  ...  \
2      0.008258256           7.31E-05      0.00195  ...   
3       0.00833959           6.04E-05      0.00176  ...   
4      0.010857733        0.000182739      0.00419  ...   
5      0.008161574        0.002668863      0.00535  ...   

0 tqwt_kurtosisValue_dec_28 tqwt_kurtosisValue_dec_29   
1                     1.562                    2.6445  \
2                    1.5589                    3.6107   
3                    1.5643                    2.3308   
4                    3.7805                    3.5664   
5                    6.1727                    5.8416   

0 tqwt_kurtosisValue_dec_30 tqwt_kurtosisValue_dec_31   
1                    3.8686                    4.2105  \
2                   23.5155                   14.1962   
3                    9.4959                   10.7458   
4                    5.2558                   14.0403   
5                    6.0805                    5.7621   

0 tqwt_kurtosisValue_dec_32 tqwt_kurtosisValue_dec_33   
1                    5.1221                    4.4625  \
2                   11.0261                    9.5082   
3                   11.0177                    4.8066   
4                    4.2235                    4.6857   
5                    7.7817                   11.6891   

0 tqwt_kurtosisValue_dec_34 tqwt_kurtosisValue_dec_35   
1                    2.6202                    3.0004  \
2                    6.5245                    6.3431   
3                    2.9199                    3.1495   
4                     4.846                     6.265   
5                    8.2103                    5.0559   

0 tqwt_kurtosisValue_dec_36 class  
1                   18.9405     1  
2                    45.178     1  
3                    4.7666     1  
4                    4.0603     1  
5                    6.1164     1  

[5 rows x 755 columns]

- Removing the duplicate data:

In [8]:
dataFrame = dataFrame.drop_duplicates()
print(f"The number of duplicate data: {dataFrame.duplicated().sum()}")

The number of duplicate data: 0


### Data Standardization:

- Removing the 'gender' and 'class' column before scaling so that their values are not affected:

In [9]:
gender = dataFrame['gender']
classData = dataFrame['class']
newDataFrame = dataFrame.drop(columns=['gender', 'class'])
newDataFrame.head(10)

0  id      PPE      DFA     RPDE numPulses numPeriodsPulses meanPeriodPulses   
1   0  0.85247  0.71826  0.57227       240              239       0.00806353  \
2   0  0.76686  0.69481  0.53966       234              233      0.008258256   
3   0  0.85083  0.67604  0.58982       232              231       0.00833959   
4   1  0.41121  0.79672  0.59257       178              177      0.010857733   
5   1   0.3279  0.79782  0.53028       236              235      0.008161574   
6   1   0.5078  0.78744  0.65451       226              221      0.007631204   
7   2  0.76095  0.62145  0.54543       322              321      0.005990989   
8   2  0.83671  0.62079  0.51179       318              317      0.006073855   
9   2  0.80826  0.61766  0.50447       318              317      0.006057188   
10  3  0.85302  0.62247  0.54855       493              492      0.003910221   

0  stdDevPeriodPulses locPctJitter locAbsJitter  ...   
1            8.68E-05      0.00218     1.76E-05  ...  \
2            7.31E-05      0.00195     1.61E-05  ...   
3            6.04E-05      0.00176     1.47E-05  ...   
4         0.000182739      0.00419     4.55E-05  ...   
5         0.002668863      0.00535     4.37E-05  ...   
6         0.002696381      0.00783     5.97E-05  ...   
7         0.000107266      0.00222     1.33E-05  ...   
8         0.000135739      0.00282     1.71E-05  ...   
9            6.93E-05      0.00161     9.73E-06  ...   
10           3.99E-05      0.00075     2.93E-06  ...   

0  tqwt_kurtosisValue_dec_27 tqwt_kurtosisValue_dec_28   
1                     1.5466                     1.562  \
2                      1.553                    1.5589   
3                     1.5399                    1.5643   
4                     6.9761                    3.7805   
5                     7.8832                    6.1727   
6                     6.2888                    4.8025   
7                     1.5541                  117.2678   
8                       1.58                    3.8564   
9                     1.6118                     2.264   
10                    2.6796                    1.6796   

0  tqwt_kurtosisValue_dec_29 tqwt_kurtosisValue_dec_30   
1                     2.6445                    3.8686  \
2                     3.6107                   23.5155   
3                     2.3308                    9.4959   
4                     3.5664                    5.2558   
5                     5.8416                    6.0805   
6                     5.0734                    7.0166   
7                    75.3156                   32.0478   
8                    11.8909                    7.2891   
9                     6.3993                    4.4165   
10                    2.0474                    2.8117   

0  tqwt_kurtosisValue_dec_31 tqwt_kurtosisValue_dec_32   
1                     4.2105                    5.1221  \
2                    14.1962                   11.0261   
3                    10.7458                   11.0177   
4                    14.0403                    4.2235   
5                     5.7621                    7.7817   
6                     5.9966                    5.2065   
7                      7.706                     3.106   
8                     4.3682                    3.6443   
9                     4.2662                    3.6357   
10                     3.507                    3.2727   

0  tqwt_kurtosisValue_dec_33 tqwt_kurtosisValue_dec_34   
1                     4.4625                    2.6202  \
2                     9.5082                    6.5245   
3                     4.8066                    2.9199   
4                     4.6857                     4.846   
5                    11.6891                    8.2103   
6                     7.4246                    3.4153   
7                     4.6206                   12.8353   
8                      5.961                   11.7552   
9                     3.7346                    2.9394   
10   

- Standardizing the remaining data:

In [10]:
scaler = StandardScaler()
standardData = pd.DataFrame(scaler.fit_transform(newDataFrame), columns=newDataFrame.columns)

- Adding the 'gender' and 'class' columns to the standardized data:

In [11]:
standardData['gender'] = gender.values
standardData['class'] = classData.values
standardData.head(10)

0        id       PPE       DFA      RPDE  numPulses  numPeriodsPulses   
0 -1.727354  0.627605  0.254241  0.607415  -0.846372         -0.841852  \
1 -1.727354  0.121892 -0.082731  0.369956  -0.906844         -0.902212   
2 -1.727354  0.617917 -0.352452  0.735211  -0.927001         -0.922333   
3 -1.713603 -1.978992  1.381694  0.755236  -1.471252         -1.465579   
4 -1.713603 -2.471119  1.397501  0.301653  -0.886687         -0.882092   
5 -1.713603 -1.408419  1.248343  1.206270  -0.987474         -1.022934   
6 -1.699852  0.086981 -1.136898  0.411972  -0.019917         -0.016922   
7 -1.699852  0.534508 -1.146382  0.167012  -0.060232         -0.057163   
8 -1.699852  0.366449 -1.191360  0.113710  -0.060232         -0.057163   
9 -1.686101  0.630853 -1.122241  0.434691   1.703543          1.703357   

0  meanPeriodPulses  stdDevPeriodPulses  locPctJitter  locAbsJitter  ...   
0          0.932476           -0.407447     -0.055085      0.037568  ...  \
1          1.039095           -0.426276     -0.142605     -0.027931  ...   
2          1.083628           -0.443731     -0.214904     -0.089065  ...   
3          2.462389           -0.275590      0.709759      1.255867  ...   
4          0.986158            3.141307      1.151162      1.177267  ...   
5          0.695764            3.179128      2.094851      1.875932  ...   
6         -0.202304           -0.379319     -0.039865     -0.150198  ...   
7         -0.156932           -0.340186      0.188447      0.015735  ...   
8         -0.166058           -0.431499     -0.271982     -0.306088  ...   
9         -1.341589           -0.471906     -0.599229     -0.603020  ...   

0  tqwt_kurtosisValue_dec_29  tqwt_kurtosisValue_dec_30   
0                  -0.619902                  -0.577334  \
1                  -0.590282                   0.192194   
2                  -0.629519                  -0.356924   
3                  -0.591640                  -0.523000   
4                  -0.521890                  -0.490698   
5                  -0.545440                  -0.454033   
6                   1.607956                   0.526386   
7                  -0.336438                  -0.443360   
8                  -0.504792                  -0.555874   
9                  -0.638207                  -0.618730   

0  tqwt_kurtosisValue_dec_31  tqwt_kurtosisValue_dec_32   
0                  -0.482710                  -0.399709  \
1                   0.015511                  -0.067675   
2                  -0.156641                  -0.068147   
3                   0.007733                  -0.450245   
4                  -0.405295                  -0.250136   
5                  -0.393595                  -0.394962   
6                  -0.308307                  -0.513092   
7                  -0.474842                  -0.482819   
8                  -0.479931                  -0.483302   
9                  -0.517810                  -0.503717   

0  tqwt_kurtosisValue_dec_33  tqwt_kurtosisValue_dec_34   
0                  -0.485012                  -0.775883  \
1                  -0.176190                  -0.527468   
2                  -0.463951                  -0.756815   
3                  -0.471351                  -0.634264   
4                  -0.042709                  -0.420207   
5                  -0.303717                  -0.725294   
6                  -0.475335                  -0.125936   
7                  -0.393296                  -0.194659   
8                  -0.529563                  -0.755574   
9                  -0.523020                  -0.691661   

0  tqwt_kurtosisValue_dec_35  tqwt_kurtosisValue_dec_36  gender  class  
0                  -0.815509                  -0.367276       1      1  
1                  -0.583818                   0.399406       1      1  
2                  -0.805174                  -0.781450       1      1  
3                  -0.589232                  -0.802089       0      1  
4                  -0.673037     

### Model Training:

- Splitting the data into training dataset(80%) and testing dataset(20%):

In [12]:
X = standardData.drop(columns=['class'])
y = standardData['class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=34)

- Training the model using Logistic Regression:

In [13]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

### Measuring Model's accuracy:

In [14]:
accuracy = accuracy_score(y_test, model.predict(X_test))
print(f"Accuracy score of the Logistic Regression Model: {accuracy * 100} %")

Accuracy score of the Logistic Regression Model: 83.44370860927152 %


- Build a predictive system to verify that the model built is able to predict the patient’s Parkinson’s disease with one instance of the dataset:

In [17]:
n = X_test.shape[0]
randomindex = np.random.randint(0, n)
result = model.predict(pd.DataFrame(X_test.iloc[randomindex, :]).T)
if result == 0:
    print("The model prediction for 1st instance:", result)
    print("Person does not have Parkinson's disease.")
else:
    print("The model prediction for 1st instance:", result)
    print("The person has Parkinson's disease.")

The model prediction for 1st instance: ['1']
The person has Parkinson's disease.
